src/InstrumentationEngine/ProfilerManager.cpp

```diff
@@ -419,6 +418,203 @@ DWORD WINAPI CProfilerManager::InstrumentationMethodThreadProc(
+                        {
+                            unordered_map<tstring, tstring> settingsMap;
+                            IfFailRet(ParseSettingsConfigurationNode(pInstrumentationMethodChildNode, settingsMap));
+
+                            for (unordered_map<tstring, tstring>::iterator it = settingsMap.begin();
+                                 it != settingsMap.end();
+                                 it++)
+                            {
+                                IfFailRet(pSource->AddSetting(it->first.c_str(), it->second.c_str()));
+                            }
+                        }
+
+                        CXmlNode* nextInstrumentationMethod = pInstrumentationMethodChildNode->Next();
+                        pInstrumentationMethodChildNode.Release();
+                        pInstrumentationMethodChildNode.Attach(nextInstrumentationMethod);
+                    }
+
+                    pProfilerManager->m_configSources.push_back(pSource.p);
+                }
+
+                CXmlNode* next = pInstrumentationMethodNode->Next();
+                pInstrumentationMethodNode.Release();
+                pInstrumentationMethodNode.Attach(next);
+            }
+        }
+        else if (wcscmp(bstrCurrChildNodeName, _T("Profilers")) == 0)
+        {
+            // TODO: handle raw profiler
+        }
+
+        CXmlNode* next = pCurrChildNode->Next();
+        pCurrChildNode.Release();
+        pCurrChildNode.Attach(next);
+    }
+
+#ifndef PLATFORM_UNIX
+    CoUninitialize();
+#endif
+
+    CLogging::LogMessage(_T("End CConfigurationLocator::ParseAttachConfigurationThreadProc"));
+
+    return 0;
+}
+
+//static
+HRESULT CProfilerManager::ParseSettingsConfigurationNode(
+    _In_ const CComPtr<CXmlNode>& parentNode,
+    _Inout_ unordered_map<tstring, tstring>& settings)
+{
+    HRESULT hr = S_OK;
+    CComPtr<CXmlNode> pSettingNode;
+    IfFailRet(parentNode->GetChildNode(&pSettingNode));
+    while (pSettingNode != nullptr)
+    {
+        CComBSTR bstrSettingNodeName;
+        IfFailRet(pSettingNode->GetName(&bstrSettingNodeName));
+
+        if (wcscmp(bstrSettingNodeName, _T("Setting")) != 0)
+        {
+            // Skip this node.
+            continue;
+        }
+
+        CComBSTR bstrSettingName;
+        IfFailRet(pSettingNode->GetAttribute(_T("Name"), &bstrSettingName));
+        IfFalseRet(bstrSettingName.Length() != 0, E_FAIL);
+
+        CComBSTR bstrSettingValue;
+        IfFailRet(pSettingNode->GetAttribute(_T("Value"), &bstrSettingValue));
+        IfFalseRet(bstrSettingValue.Length() != 0, E_FAIL);
+
+        if (settings.find(bstrSettingName.m_str) == settings.end())
+        {
+            settings.insert(std::pair<tstring, tstring>(bstrSettingName, bstrSettingValue));
+        }
+        else
+        {
+            // Override value.
+            settings[bstrSettingName.m_str] = bstrSettingValue;
+        }
```

Comment:
Is there any value in allowing a setting to be declared more than one time? I think we should start out by failing if a setting appears more than once. We can relax that later if it's somehow useful to allow for this to occur.

Reply:
I was mirroring the InstrumentationEngine logic which does allow duplicates due to parsing in a loop, but I agree with your point.

---
In reply to: [386707618](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386707618) [](ancestors = 386707618)

Reply:
Removed the else statement

---
In reply to: [386708529](https://github.com/microsoft/CLRInstrumentationEngine/pull/216#discussion_r386708529) [](ancestors = 386708529,386707618)